File: kits/zig/test/truth_host_test.zig
//! Part three of `papers/quo-truth.md`: what the host does. The same being,
//! unchanged, is installed under a warden reached by every road this kit has
//! and gives the same answers; a ground that publishes nothing is pushed to
//! down the line it holds; a closed one is weather. Written from that part
//! alone.

const std = @import("std");
const wire = @import("wire");
const warden = @import("warden");
const quo = @import("quo");
const host = @import("host");

const Key = warden.Key;

/// A clock that moves, because the roads here are real and an ask that waits
/// for an answer waits against a budget. It is handed to the host, which
/// hands it to the warden; nothing below reaches for one.
var ticking: std.atomic.Value(i64) = .init(1_000);

fn now() i64 {
    return ticking.fetchAdd(1, .acq_rel);
}

var draws: std.atomic.Value(u64) = .init(0);

fn random() Key {
    var out: Key = undefined;
    var seed = draws.fetchAdd(1, .acq_rel) *% 0x9e3779b97f4a7c15 +% @as(u64, 0x51_7c_c1);
    for (&out) |*byte| {
        seed = seed *% 6364136223846793005 +% 1442695040888963407;
        byte.* = @truncate(seed >> 33);
    }
    return out;
}

fn founding() warden.Warden.Seeds {
    return .{ .name = random(), .padlock = random(), .heir = random() };
}

const DOG =
    \\Dog
    \\  name() text
    \\  logWalk(minutes int) bool
    \\
;

const Dog = struct {
    quo: quo.Cell = .{},
    gpa: std.mem.Allocator,
    walks: std.ArrayList(i64) = .empty,

    pub fn deinit(self: *Dog) void {
        self.walks.deinit(self.gpa);
    }

    pub fn name(_: *Dog) []const u8 {
        return "Rex";
    }

    pub fn logWalk(self: *Dog, minutes: i64) bool {
        self.walks.append(self.gpa, minutes) catch return false;
        return true;
    }
};

const INBOX =
    \\Inbox
    \\  walked(minutes int)
    \\
;

const Inbox = struct {
    quo: quo.Cell = .{},
    gpa: std.mem.Allocator,
    heard: std.ArrayList(i64) = .empty,

    pub fn deinit(self: *Inbox) void {
        self.heard.deinit(self.gpa);
    }

    pub fn walked(self: *Inbox, minutes: i64) void {
        self.heard.append(self.gpa, minutes) catch {};
    }
};

const WALKER =
    \\Walker
    \\  subscribe(inbox invitation) bool
    \\  walk(minutes int) bool
    \\
;

const Walker = struct {
    quo: quo.Cell = .{},

    pub fn subscribe(_: *Walker, at: *quo.At, inbox: wire.Invitation) bool {
        var accepted = (at.accept(inbox) catch return false) orelse return false;
        defer accepted.deinit();
        const listener = accepted.only() orelse return false;
        at.label("inbox", listener) catch return false;
        return true;
    }

    pub fn walk(_: *Walker, at: *quo.At, minutes: i64) bool {
        const inbox = at.relation("inbox") orelse return true;
        _ = inbox.call(at.gpa, void, "walked", .{minutes}) catch {};
        return true;
    }
};

fn stand(gpa: std.mem.Allocator, io: std.Io, roads: []const host.Roads, at: *host.Host) !void {
    return host.Host.open(gpa, .{
        .seeds = founding(),
        .clock = now,
        .random = random,
        .io = io,
        .roads = roads,
    }, at);
}

fn walkFrom(door: *warden.Warden, gpa: std.mem.Allocator) quo.At {
    return .{ .door = door, .being = door.name, .gpa = gpa };
}

/// Accept an invitation that opens one being, under a label of this ground's
/// own.
fn joining(
    door: *warden.Warden,
    gpa: std.mem.Allocator,
    invitation: wire.Invitation,
    label: []const u8,
) !?quo.Handle {
    var accepted = (try quo.accepting(door, gpa, invitation, walkFrom(door, gpa))) orelse
        return null;
    defer accepted.deinit();
    const handle = accepted.only() orelse return null;
    try walkFrom(door, gpa).label(label, handle);
    return handle;
}

// -------------------------------------------------- the same being, any road

test "the same Dog, installed behind every road this kit has, gives the same answers" {
    const gpa = std.testing.allocator;
    var threaded: std.Io.Threaded = .init(gpa, .{});
    defer threaded.deinit();
    const io = threaded.io();

    for ([_]host.Roads{ .memory, .http, .line }) |road| {
        var alice: host.Host = undefined;
        try stand(gpa, io, &.{road}, &alice);
        defer alice.close();
        var bob: host.Host = undefined;
        try stand(gpa, io, &.{road}, &bob);
        defer bob.close();

        var rex: Dog = .{ .gpa = gpa };
        defer rex.deinit();
        const held = try quo.holding(&alice.door, Dog, &rex, DOG, .{}, gpa);

        const invitation = try held.under.grant(gpa, null);
        defer gpa.free(invitation.hints);
        const handle = (try joining(&bob.door, gpa, invitation, "rex")) orelse
            return error.TestUnexpectedResult;

        const said = (try handle.call(gpa, []const u8, "name", .{})) orelse
            return error.TestUnexpectedResult;
        defer gpa.free(said);
        try std.testing.expectEqualStrings("Rex", said);
        try std.testing.expectEqual(
            @as(?bool, true),
            try handle.call(gpa, bool, "logWalk", .{@as(i64, 12)}),
        );
        try std.testing.expectEqualSlices(i64, &.{12}, rex.walks.items);

        // The being never learned the road. `quo.At` has no field naming one,
        // and neither has the cell the being holds.
        inline for (@typeInfo(quo.Cell).@"struct".fields) |field| {
            try std.testing.expect(std.mem.indexOf(u8, field.name, "road") == null);
            try std.testing.expect(std.mem.indexOf(u8, field.name, "hint") == null);
        }
    }
}

test "a hint the caller cannot speak is walked past, and the road it can speak carries" {
    const gpa = std.testing.allocator;
    var threaded: std.Io.Threaded = .init(gpa, .{});
    defer threaded.deinit();
    const io = threaded.io();

    // Alice publishes a road nobody here can speak, first, and one it can
    // after it.
    var alice: host.Host = undefined;
    try host.Host.open(gpa, .{
        .seeds = founding(),
        .clock = now,
        .random = random,
        .io = io,
        .roads = &.{.memory},
        .hints = &.{"pigeon://loft"},
    }, &alice);
    defer alice.close();
    var bob: host.Host = undefined;
    try stand(gpa, io, &.{.memory}, &bob);
    defer bob.close();

    var rex: Dog = .{ .gpa = gpa };
    defer rex.deinit();
    const held = try quo.holding(&alice.door, Dog, &rex, DOG, .{}, gpa);
    const invitation = try held.under.grant(gpa, null);
    defer gpa.free(invitation.hints);
    try std.testing.expectEqual(@as(usize, 2), invitation.hints.len);
    try std.testing.expectEqualStrings("pigeon://loft", invitation.hints[0]);

    const handle = (try joining(&bob.door, gpa, invitation, "rex")) orelse
        return error.TestUnexpectedResult;
    const said = (try handle.call(gpa, []const u8, "name", .{})) orelse
        return error.TestUnexpectedResult;
    defer gpa.free(said);
    try std.testing.expectEqualStrings("Rex", said);
}

test "a ground that publishes nothing: its pushes ride back down the line it holds" {
    const gpa = std.testing.allocator;
    var threaded: std.Io.Threaded = .init(gpa, .{});
    defer threaded.deinit();
    const io = threaded.io();

    // Bob's laptop listens. Alice's tab has no road of its own and dials out.
    var laptop: host.Host = undefined;
    try stand(gpa, io, &.{.line}, &laptop);
    defer laptop.close();
    var tab: host.Host = undefined;
    try stand(gpa, io, &.{}, &tab);
    defer tab.close();
    try std.testing.expectEqual(@as(usize, 0), tab.door.hints.items.len);

    var walker: Walker = .{};
    var inbox: Inbox = .{ .gpa = gpa };
    defer inbox.deinit();
    const at_walker = try quo.holding(&laptop.door, Walker, &walker, WALKER, .{}, gpa);
    const at_inbox = try quo.holding(&tab.door, Inbox, &inbox, INBOX, .{}, gpa);

    const to_walker = try at_walker.under.grant(gpa, null);
    defer gpa.free(to_walker.hints);
    const bob = (try joining(&tab.door, gpa, to_walker, "walker")) orelse
        return error.TestUnexpectedResult;

    const to_inbox = try at_inbox.under.grant(gpa, null);
    defer gpa.free(to_inbox.hints);
    // The tab publishes nothing, so the invitation carries no road at all.
    try std.testing.expectEqual(@as(usize, 0), to_inbox.hints.len);

    try std.testing.expectEqual(
        @as(?bool, true),
        try bob.call(gpa, bool, "subscribe", .{to_inbox}),
    );
    try std.testing.expectEqual(@as(?bool, true), try bob.call(gpa, bool, "walk", .{@as(i64, 9)}));
    try std.testing.expectEqual(@as(?bool, true), try bob.call(gpa, bool, "walk", .{@as(i64, 11)}));
    try std.testing.expectEqualSlices(i64, &.{ 9, 11 }, inbox.heard.items);
}

test "a closed line is weather: the push meets silence, the number is spent, nothing throws" {
    const gpa = std.testing.allocator;
    var threaded: std.Io.Threaded = .init(gpa, .{});
    defer threaded.deinit();
    const io = threaded.io();

    var laptop: host.Host = undefined;
    try stand(gpa, io, &.{.line}, &laptop);
    defer laptop.close();

    var walker: Walker = .{};
    const at_walker = try quo.holding(&laptop.door, Walker, &walker, WALKER, .{}, gpa);

    var inbox: Inbox = .{ .gpa = gpa };
    defer inbox.deinit();
    {
        var tab: host.Host = undefined;
        try stand(gpa, io, &.{}, &tab);
        defer tab.close();

        const at_inbox = try quo.holding(&tab.door, Inbox, &inbox, INBOX, .{}, gpa);
        const to_walker = try at_walker.under.grant(gpa, null);
        defer gpa.free(to_walker.hints);
        const bob = (try joining(&tab.door, gpa, to_walker, "walker")) orelse
            return error.TestUnexpectedResult;
        const to_inbox = try at_inbox.under.grant(gpa, null);
        defer gpa.free(to_inbox.hints);
        _ = try bob.call(gpa, bool, "subscribe", .{to_inbox});
        _ = try bob.call(gpa, bool, "walk", .{@as(i64, 1)});
    }

    // The tab is gone. Walker's own answer to itself is unaffected; only the
    // push found nobody, and the source cannot tell that from a released
    // being.
    try std.testing.expectEqual(
        @as(?bool, true),
        try at_walker.call(gpa, bool, "walk", .{@as(i64, 2)}),
    );
    try std.testing.expectEqualSlices(i64, &.{1}, inbox.heard.items);
}

test "a road never holds a secret: what the host keeps per peer is an address" {
    // The host's own table is a padlock beside a line, and a padlock is a
    // public key used as an address — the only kind of key a road ever sees.
    // What the warden hands downward is that address and an opaque token, and
    // there is nowhere in the shape of the call to put anything else.
    const arrived = @typeInfo(@FieldType(warden.Delivery, "arrived")).optional.child;
    const params = @typeInfo(@typeInfo(arrived).pointer.child).@"fn".params;
    try std.testing.expectEqual(@as(usize, 3), params.len);
    try std.testing.expectEqual(Key, params[1].type.?);
    try std.testing.expectEqual(?*anyopaque, params[2].type.?);
}

File: kits/js/bench/truth-host.test.js
// Part three of papers/quo-truth.md: what the host does. The same being,
// unchanged, is installed under a warden reached by four roads and gives the
// same answers; a tab that publishes nothing is pushed to down the line it
// holds; a closed tab is weather. Written from the paper alone.
import test from 'node:test';
import assert from 'node:assert/strict';
import { MemoryStore, Warden, memoryDelivery, seeds } from '../src/index.js';
import { host } from '../src/host.js';
import { TLS, TRUST } from './tls.js';

const still = () => 1_000;
const random = () => crypto.getRandomValues(new Uint8Array(32));

const DOG = `Dog
  name() text
  logWalk(minutes int) bool
`;

class Dog {
  constructor() {
    this.walks = [];
  }
  name() {
    return 'Rex';
  }
  logWalk(minutes) {
    this.walks.push(minutes);
    return true;
  }
}

const INBOX = `Inbox
  walked(minutes int)
`;

class Inbox {
  constructor() {
    this.heard = [];
  }
  walked(minutes) {
    this.heard.push(minutes);
  }
}

const WALKER = `Walker
  subscribe(inbox invitation) bool
  walk(minutes int) bool
`;

class Walker {
  async subscribe(invitation) {
    this.listener = (await this.quo.accept(invitation, { label: 'inbox' }))?.[0] ?? null;
    return this.listener !== null;
  }
  async walk(minutes) {
    await this.listener?.walked(minutes);
    return true;
  }
}

// The host: seeds, clock, randomness, store and delivery handed in, roads
// stood up in front of the warden's one entry point. `host` is the kit's own
// hosting for Node, and the roads it can stand are the ones the law names.
async function stand(roads, hints = []) {
  return host({ seeds: seeds(random), clock: still, random, roads, hints, tls: TLS, trust: TRUST });
}

for (const roads of [['http'], ['tcp'], ['wss'], ['memory']]) {
  test(`the same Dog, installed behind ${roads[0]}, gives the same answers`, async () => {
    const alice = await stand(roads);
    const bob = await stand(roads);
    try {
      const rex = new Dog();
      await alice.warden.hold(rex, { blueprint: DOG });
      const [handle] = await bob.warden.accept(await rex.quo.grant(rex), { label: 'rex' });
      assert.equal(await handle.name(), 'Rex');
      assert.equal(await handle.logWalk(12n), true);
      assert.deepEqual(rex.walks, [12n]);
      // The being never learned the road: nothing on it names one.
      assert.equal(rex.quo.road, undefined);
    } finally {
      await alice.close();
      await bob.close();
    }
  });
}

test('a hint the caller cannot speak is walked past, and the road it can speak carries', async () => {
  // Alice publishes a road nobody here can speak, first, and HTTP after it.
  const alice = await stand(['http'], ['pigeon://loft']);
  const bob = await stand(['http']);
  try {
    const rex = new Dog();
    await alice.warden.hold(rex, { blueprint: DOG });
    const invitation = await rex.quo.grant(rex);
    assert.deepEqual(invitation.hints.length, 2);
    assert.equal(invitation.hints[0], 'pigeon://loft');
    const [handle] = await bob.warden.accept(invitation, { label: 'rex' });
    assert.equal(await handle.name(), 'Rex');
  } finally {
    await alice.close();
    await bob.close();
  }
});

test('a tab publishes nothing: its pushes ride back down the line it holds', async () => {
  // Bob's laptop listens. Alice's tab has no road of its own and dials out.
  const laptop = await stand(['wss']);
  const tab = await stand([]);
  try {
    const walker = new Walker();
    const inbox = new Inbox();
    await laptop.warden.hold(walker, { blueprint: WALKER });
    await tab.warden.hold(inbox, { blueprint: INBOX });
    assert.deepEqual(tab.warden.hints, []);

    const [bob] = await inbox.quo.accept(await walker.quo.grant(walker), { label: 'walker' });
    assert.equal(await bob.subscribe(await inbox.quo.grant(inbox)), true);
    await bob.walk(9n);
    await bob.walk(11n);
    assert.deepEqual(inbox.heard, [9n, 11n]);
  } finally {
    await laptop.close();
    await tab.close();
  }
});

test('a closed tab is weather: the push meets silence, the number is spent, nothing throws', async () => {
  const laptop = await stand(['tcp']);
  const tab = await stand([]);
  const walker = new Walker();
  const inbox = new Inbox();
  await laptop.warden.hold(walker, { blueprint: WALKER });
  await tab.warden.hold(inbox, { blueprint: INBOX });
  const [bob] = await inbox.quo.accept(await walker.quo.grant(walker), { label: 'walker' });
  await bob.subscribe(await inbox.quo.grant(inbox));
  await bob.walk(1n);
  await tab.close();
  // Walker's own answer to itself is unaffected; only the push found nobody.
  assert.equal(await walker.walk(2n), true);
  assert.deepEqual(inbox.heard, [1n]);
  await laptop.close();
});

test('a tab that closes and opens again is answered on its first ask', async () => {
  // The store is written on the acts that change it and not on every ask, so a
  // ground that stops between two asks comes back holding a number the far
  // door has already spent — and its first ask back is refused as the replay
  // it looks exactly like, after the caller has waited out its whole
  // allowance. A ground that knows it is stopping writes its records first,
  // and `close` is where it knows.
  const laptop = await stand(['tcp']);
  const walker = new Walker();
  await laptop.warden.hold(walker, { blueprint: WALKER });

  const store = new MemoryStore();
  const kept = { name: random(), padlock: random(), heir: random() };
  const seed = random();

  const open = async () => {
    const tab = await host({ seeds: kept, clock: still, random, roads: [], store });
    const inbox = new Inbox();
    await tab.warden.hold(inbox, { seed, blueprint: INBOX });
    return { tab, inbox };
  };

  const first = await open();
  const [bob] = await first.inbox.quo.accept(await walker.quo.grant(walker), { label: 'walker' });
  // Two asks after the last act that owed a write, so the number is two ahead
  // of what the store would otherwise have held.
  assert.equal(await bob.walk(1n), true);
  assert.equal(await bob.walk(2n), true);
  await first.tab.close();

  const again = await open();
  try {
    const back = again.tab.warden.relation('walker');
    assert.ok(back, 'the relation survived the tab');
    assert.equal(await back.walk(3n), true, 'and the first ask back was answered, not refused');
  } finally {
    await again.tab.close();
    await laptop.close();
  }
});

test('a road never holds a secret: what the host is handed is seeds, and what it keeps is addresses', async () => {
  const delivery = memoryDelivery();
  const warden = await Warden.open({ seeds: seeds(random), clock: still, random, delivery });
  delivery.attach('mem://one', warden);
  // What delivery is given per row is the way back and nothing else.
  const rows = [];
  delivery.watch((row) => rows.push(row));
  const rex = new Dog();
  await warden.hold(rex, { blueprint: DOG });
  const other = await Warden.open({ seeds: seeds(random), clock: still, random, delivery });
  delivery.attach('mem://two', other);
  other.publish('mem://two');
  warden.publish('mem://one');
  const [handle] = await other.accept(await rex.quo.grant(rex), { label: 'rex' });
  await handle.name();
  assert.ok(rows.length > 0);
  for (const row of rows) {
    assert.deepEqual(Object.keys(row).sort(), ['hints', 'padlock']);
  }
});
